src/CanvasContainer.tsx

```diff
@@ -2,171 +2,400 @@ import React, { CSSProperties, useEffect, useRef } from 'react';
 import { canvasModel, ZoomFactor } from './canvasMachine';
 import { useCanvas } from './CanvasContext';
 import { useMachine } from '@xstate/react';
+import {
+  assign,
+  actions,
+  sendParent,
+  ContextFrom,
+  SpecialTargets,
+} from 'xstate';
 import { createModel } from 'xstate/lib/model';
 import { Point } from './pathUtils';
-import { AnyState, EmbedContext } from './types';
+import { isAcceptingSpaceNatively, isWithPlatformMetaKey } from './utils';
 import { useEmbed } from './embedContext';
-import { isWithPlatformMetaKey, isTextInputLikeElement } from './utils';
+import { AnyState } from './types';
 
-const dragModel = createModel(
+interface DragSession {
+  pointerId: number;
+  point: Point;
+}
+
+interface PointDelta {
+  x: number;
+  y: number;
+}
+
+const dragSessionModel = createModel(
   {
-    startPoint: { x: 0, y: 0 },
-    dragPoint: { x: 0, y: 0 },
-    dx: 0,
-    dy: 0,
-    embed: undefined as EmbedContext | undefined,
+    session: null as DragSession | null,
+    ref: null as React.MutableRefObject<HTMLElement> | null,
```

Comment:
We might need a better naming to what this ref refers.

Reply:
Also might be better to just save the HTML element in the context instead of tightening the model to React.

Reply:
> We might need a better naming to what this ref refers to.

From the point of this machine, it's an "abstract" ref - it doesn't refer to any particular in-app ref as the machine is generalized (I think it could actually be reused with success and could replace the one defined in the [ResizableBox](https://github.com/statelyai/xstate-viz/blob/c36ce2329c8bf0a900134be03733b27ad3502f60/src/ResizableBox.tsx#L23)).

The `ref` usually refers to a DOM ref (if it's not using some other name) so I've assumed that using this here would be OK. Would you prefer something like `domRef`?

> Also might be better to just save the HTML element in the context instead of tightening the model to React.

It could but I would have to sync the ref to a machine when the component mounts which seemed like a hassle and something that would complicate the implementation here.

Reply:
>It could but I would have to sync the ref to a machine when the component mounts which seemed like a hassle and something that would complicate the implementation here.
I mean just passing `ref.current` directly to the machine. we won't need to sync anything back from the machine to the component so ref in the machine is always read-only.

Reply:
@farskid that's impossible - we pass `ref` to a machine during render but `ref.current` gets assigned for the first time in the commit phase of React.

Reply:
ok got it.